libshortfin/src/shortfin/support/blocking_executor.cc

```diff
@@ -0,0 +1,162 @@
+// Copyright 2024 Advanced Micro Devices, Inc
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "shortfin/support/blocking_executor.h"
+
+#include "fmt/core.h"
+#include "shortfin/support/logging.h"
+
+namespace shortfin {
+
+BlockingExecutor::BlockingExecutor(iree_allocator_t allocator)
+    : allocator_(allocator) {}
+BlockingExecutor::~BlockingExecutor() { Kill(/*wait=*/true); }
+
+bool BlockingExecutor::has_free_threads() {
+  iree::slim_mutex_lock_guard g(control_mu_);
+  return free_threads_ != nullptr;
+}
+
+void BlockingExecutor::Kill(bool wait, iree_timeout_t warn_timeout) {
+  {
+    iree::slim_mutex_lock_guard g(control_mu_);
+    kill_ = true;
+  }
+
+  // Because there can be some asynchronicity in terms of the kill signal and
+  // additional scheduling, we loop on a short delay, sending kill signals,
+  // issuing a warning if waiting for a long time. Once the number of
+  // live threads drops to zero, we set the inhibit flag, causing no more
+  // work to be accepted and exit.
+  iree_time_t report_deadline = iree_timeout_as_deadline_ns(warn_timeout);
+  for (;;) {
+    // Wake up all free threads, leaving their task null. This will cause them
+    // to loop once and ask the executor to add them back to the free list
+    // (at which point, they will be told to shutdown).
+    {
+      iree::slim_mutex_lock_guard g(control_mu_);
+      for (;;) {
+        ThreadInstance *current = free_threads_;
+        if (!current) break;
+        current->signal_transact.set();
+        free_threads_ = current->next;
+        current->next = nullptr;
+      }
+    }
+
+    // If not asked to wait, just exit.
+    if (!wait) {
+      break;
+    }
+
+    // Check if still alive.
+    int last_live_thread_count;
+    int total_thread_count;
+    {
+      iree::slim_mutex_lock_guard g(control_mu_);
+      last_live_thread_count = live_thread_count_;
+      total_thread_count = created_thread_count_;
+      if (live_thread_count_ == 0) {
+        inhibit_ = true;
+        break;
+      }
+    }
+
+    // Report if over the reporting deadling.
+    if (iree_time_now() > report_deadline) {
+      logging::warn(
+          "Still waiting for blocking executor threads to terminate (live "
+          "threads = {}, total created = {}).",
+          last_live_thread_count, total_thread_count);
+      report_deadline = iree_timeout_as_deadline_ns(warn_timeout);
+    }
+
+    // Short spin delay to signal/wait.
+    iree_wait_until(iree_timeout_as_deadline_ns(iree_make_timeout_ms(250)));
```

Comment:
250ms is quite awhile - shouldn't be needed? you can use `iree_thread_join` to join with the thread - if the issue is that you don't want an infinite wait there we could add a iree_thread_try_join that takes a timeout

Reply:
Yeah, isn't great but I really do want the deadline (have debugged too many shutdown issues to not have hang reporting). Dropped the timeout to 50ms and added a TODO: `    // TODO: Introduce iree_thread_try_join and refactor the shutdown sequence
    // to be based on thread joining with a warn deadline.`